scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -0,0 +1,158 @@
+                                   to HostInfo schema above"))
+
+(def Schedule
+  {TimePeriodMs {:suggested-matches {HostInfo [s/Uuid]}
+                 :suggested-purchases {HostInfo NonNegInt}}})
+
+(defprotocol Optimizer
+  "Protocol defining a tool to produce a schedule to execute"
+  (produce-schedule [this queue running available host-infos]
+                    "Returns a schedule of what jobs to place on what machines
+                     and what hosts to purchase at different time steps.
+                     Conforms to the Schedule schema above
+
+                     Parameters:
+                     queue -- Ordered list of jobs to run (see cook.mesos.schema job)
+                     running -- Set of tasks running (see cook.mesos.schema instance)
+                     available -- Set of offers outstanding
+                     host-infos -- Host infos from HostFeed"))
+
+(defprotocol ScheduleConsumer
+  "Protocol defining a consumer of the optimizer's schedule.
+   An examples is a component to take the schedules suggested
+   purchases and execute on them"
+  ;;Question to self: Should this be run its own thread or let it kick off a thread
+  (consume-schedule [this schedule]
+                    "Act on the schedule passed in.
+                     It is expected this function has side effects
+                     and it will be run in a separate thread"))
+
+(defn create-dummy-host-feed
+  "Returns an instance of HostFeed which returns an empty list of hosts"
+  [_]
+  (reify HostFeed
+    (get-available-host-info [this]
+      [])))
+
+(defn create-dummy-optimizer
+  "Returns an instance of Optimizer which returns an empty schedule"
+  [_]
+  (reify Optimizer
+    (produce-schedule [this queue running available host-infos]
+      {0 {:suggested-purchases {} :suggested-matches {}}})))
+
+(defn create-dummy-consumer
+  "Returns an instance of ScheduleConsumer which does nothing"
+  [_]
+  (reify ScheduleConsumer
+    (consume-schedule [this schedule])))
+
+(defn optimizer-cycle!
+  "Starts a cycle that:
+   1. Gets queue, running, offer and purchasable host info
+   2. Calls the `optimizer` to get a schedule
+   3. Passes the schedule to each of the consumers
+
+   Parameters:
+   get-queue -- fn, no args fn that returns ordered list of jobs to run
+   get-running -- fn, no args fn that returns a set of tasks running
+   get-offers -- fn, no args fn that returns a set of offers
+   host-feed -- instance of HostFeed
+   optimizer -- instance of Optimizer
+   schedule-consumer -- instance of ScheduleConsumer
+   interval -- joda-time period
+
+   Raises an exception if there was a problem in the execution"
+  [get-queue get-running get-offers host-feed optimizer schedule-consumer]
+  (let [queue (future (get-queue))
+        running (future (get-running))
+        offers (future (get-offers))
+        host-infos (future (get-available-host-info host-feed))
+        _ (s/validate [HostInfo] @host-infos)
```

Comment:
Isn't the future for `host-infos` wasteful as we immediately try to realize the value?

Reply:
whoops, missed this one. Give me a second